data_model/src/block.rs

```diff
@@ -10,274 +10,281 @@ use core::{cmp::Ordering, fmt::Display};
+use crate::{events::prelude::*, peer, transaction::prelude::*};
 
-mod header {
-    pub use self::model::*;
+/// Trait for basic block operations
+pub trait Block {
+    /// Calculate block hash
+    #[cfg(feature = "std")]
+    fn hash(&self) -> HashOf<BlockPayload> {
+        HashOf::new(self.header()).transmute()
+    }
+    /// Return block header
+    fn header(&self) -> &BlockHeader {
+        &self.payload().header
+    }
+
+    /// Return block payload
+    fn payload(&self) -> &BlockPayload;
+    /// Return block signatures
+    fn signatures(&self) -> &SignaturesOf<BlockPayload>;
+}
+
+#[model]
+pub mod model {
     use super::*;
+    use crate::transaction::error::TransactionRejectionReason;
+
+    #[derive(
+        Debug,
+        Display,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        Getters,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[cfg_attr(
+        feature = "std",
+        display(fmt = "Block №{height} (hash: {});", "HashOf::new(&self)")
+    )]
+    #[cfg_attr(not(feature = "std"), display(fmt = "Block №{height}"))]
+    #[getset(get = "pub")]
+    #[allow(missing_docs)]
+    #[ffi_type]
+    // TODO: Do we need both BlockPayload and BlockHeader?
+    // If yes, what data goes into which structure?
+    pub struct BlockHeader {
+        /// Number of blocks in the chain including this block.
+        pub height: u64,
+        /// Creation timestamp (unix time in milliseconds).
+        pub timestamp_ms: u128,
+        /// Hash of the previous block in the chain.
+        pub previous_block_hash: Option<HashOf<BlockPayload>>,
+        /// Hash of merkle tree root of valid transactions' hashes.
+        pub transactions_hash: Option<HashOf<MerkleTree<TransactionPayload>>>,
+        /// Hash of merkle tree root of rejected transactions' hashes.
+        pub rejected_transactions_hash: Option<HashOf<MerkleTree<TransactionPayload>>>,
+        /// Topology of the network at the time of block commit.
+        pub commit_topology: Vec<peer::PeerId>,
+        /// Value of view change index. Used to resolve soft forks.
+        // NOTE: This field used to be required to rotate topology. After merging
+        // https://github.com/hyperledger/iroha/pull/3250 only commit_topology is used
+        #[deprecated(since = "2.0.0-pre-rc.13", note = "Will be removed in future versions")]
+        pub view_change_index: u64,
+        /// Estimation of consensus duration (in milliseconds).
+        pub consensus_estimation_ms: u64,
+    }
 
-    #[model]
-    pub mod model {
-        use super::*;
+    #[derive(
+        Debug, Display, Clone, Eq, Getters, Decode, Encode, Deserialize, Serialize, IntoSchema,
+    )]
+    #[display(fmt = "({header})")]
+    #[getset(get = "pub")]
+    #[allow(missing_docs)]
+    #[ffi_type]
+    pub struct BlockPayload {
+        /// Block header
```

Comment:
What is the motivation for this + the block trait? Seems more complicated in my opinion.